File: libs/redis/src/schemas/command.ts
import { Command as CommandType, CommandModule, CommandPermission, CooldownType, Prisma } from '@tsuwari/prisma';
import { Entity, Schema } from 'redis-om';

type CMD = CommandType & { responses: string[] }

export class Command extends Entity implements CMD {
  id: string;
  name: string;
  cooldown: number | null;
  cooldownType: CooldownType;
  enabled: boolean;
  aliases: string[];
  description: string | null;
  visible: boolean;
  channelId: string;
  permission: CommandPermission;
  default: boolean;
  module: CommandModule;
  defaultName: string | null;
  responses: string[];
}

export const commandSchema = new Schema(Command, {
  id: { type: 'string', indexed: true },
  name: { type: 'string', indexed: true },
  cooldown: { type: 'number' },
  cooldownType: { type: 'string' },
  enabled: { type: 'boolean' },
  aliases: { type: 'string[]' },
  description: { type: 'string' },
  visible: { type: 'boolean' },
  channelId: { type: 'string' },
  permission: { type: 'string' },
  default: { type: 'boolean' },
  module: { type: 'string' },
  defaultName: { type: 'string' },
  responses: { type: 'string[]' },
}, {
  prefix: 'commands',
  indexedDefault: true,
});

File: apps/api-gql/internal/entity/command.go
package entity

import (
	"time"

	"github.com/google/uuid"
)

type Command struct {
	ID                        uuid.UUID
	Name                      string
	Cooldown                  *int
	CooldownType              string
	Enabled                   bool
	Aliases                   []string
	Description               *string
	Visible                   bool
	ChannelID                 string
	Default                   bool
	DefaultName               *string
	Module                    string
	IsReply                   bool
	KeepResponsesOrder        bool
	DeniedUsersIDS            []string
	AllowedUsersIDS           []string
	RolesIDS                  []uuid.UUID
	OnlineOnly                bool
	OfflineOnly               bool
	CooldownRolesIDs          []string
	EnabledCategories         []string
	RequiredWatchTime         int
	RequiredMessages          int
	RequiredUsedChannelPoints int
	GroupID                   *uuid.UUID
	ExpiresAt                 *time.Time
	ExpiresType               *CommandExpireType
}

var CommandNil = Command{}

type CommandExpireType string

const (
	CommandExpireTypeDisable CommandExpireType = "DISABLE"
	CommandExpireTypeDelete  CommandExpireType = "DELETE"
)

type CommandGroup struct {
	ID        uuid.UUID
	ChannelID string
	Name      string
	Color     string
}

var CommandGroupNil = CommandGroup{}

type CommandResponse struct {
	ID                uuid.UUID
	Text              *string
	CommandID         uuid.UUID
	Order             int
	TwitchCategoryIDs []string
	OnlineOnly        bool
	OfflineOnly       bool
}

var CommandResponseNil = CommandResponse{}

type CommandWithGroupAndResponses struct {
	Command   Command
	Group     *CommandGroup
	Responses []CommandResponse
}

var CommandWithGroupAndResponsesNil = CommandWithGroupAndResponses{}
